src/main/protobuf/CommonIndexFileFormat.proto

```diff
@@ -1,6 +1,11 @@
 syntax = "proto3";
 package io.osirrc.ciff;
 
+message Header {
+  int32 num_postings_lists = 1;
+  int32 num_docs = 2;
+}
```

Comment:
To get it all into a single ProtoBuff file that doesn't rely on knowing what's in there, what's repeated and how many times I suggest:

message CIFF {
  int32 version = 1;
  int32 num_postings_lists = 2;
  int32 num_docs = 3;
  repeat DocRecord = 4;
  repeat PostingsList = 5;
}

I've also renamed header to CIFF (for clarity) and added a version element.  This is so that if we later add stuff (such as fields) we can easily reject the CIFF without looking for stuff we don't know about in a large file we can parse most of.

Reply:
@andrewtrotman We can't do this because it would require CIFF to be completely held in memory, per the rationale discussed in the README.

Also, CIFF is a format, not a message, so the name isn't totally appropriate.

Reply:
How about `IndexMetadata`? I'm fine with adding a version.

Reply:
Would it be an idea to also add `Metadata` for each `docrecord`?

Reply:
What do you mean exactly? There will be a `DocRecord` message for every single document in the collection...

Reply:
Like an `optional metadata` that can include information like publication date / author in the `core18` collection.

Reply:
This should go in the metadata header... yes, we can throw in avgdl, num docs, num terms, collection name, etc.

Reply:
>it would require CIFF to be completely held in memory, per the rationale discussed in the README.

Sure, but the ClueWeb index is so small that it fits in memory anyway.  Indeed, this is how my code works, it block reads the CIFF index and converts to a JASS index all in memory, the dumps the JASS index and terminates.  Although we are trying to be future-proof with CIFF, I find it unlikely that we'll ever generate a CIFF file too large to fit into memory - because we already assume elsewhere that the search engine's index is memory resident.

Reply:
Current CW12b index is small because it only has postings for the query terms... postings lists for CW12b is 87 GiB (uncompressed) in the current protobuf format for _all_ postings lists.

Reply:
In order to compute the impact scores in JASSv2 I need the max and min BM25 score across the entire index - so I use the 87GiB version.  Come on, 87GiB is small.  Its a pig to transfer, but small by comparison to the amount of memory on a modern desktop.

Reply:
That means I couldn't do an export from my laptop, even though my laptop can comfortably hold the raw index...?

Reply:
protobuf repeated fields are BLOBs stored with the BLOB-length at the start, so (I'm guessing) you need enough memory to compute the repeated field so as to compute the length so as to write the length then the data.  But, my laptop only has 8GiB RAM, so I can't even get the ClueWeb index into RAM even once it is converted from CIFF.  So I don't really buy the argument that "my laptop isn't a supercomputer so we shouldn't do this".

I'll add at this point that I'm happy to go with the consensus on this point.  My experience, however, is that if you have a multiple file index then something will go wrong that can't go wrong with a single file - which is why ATIRE has a single file for the index.  Yes, its a bit painful, but you only pay the pain once (when writing the code to write and read indexes).